# **자연어 처리 Exp1 Report**
                                                컴퓨터전자시스템공학부 201603727 허성일
                                                제출일 : 2020-09-24
                                                교수님 : 김낙현 교수님

**[실습내용]**

1. p. 8의 프로그램을 수정하여 ‘소나기.txt’ 파일을 분석함. 나온 횟수가 5번 이상인 단어들을 출력함. 화면에 나온 데이터를 복사해서 제출해도 됨
	
2. 불용어 처리: 결과에 특수문자(. , “ ( ) 등)와 조사(‘이‘, ‘은‘, ‘다‘, …)가 포함되지 않도록 불용어 리스트를 만들고, 이 리스트가 적용된 결과를 다시 구함. 나온 횟수가 5번 이상인 단어들을 출력함. 불용어 리스트는 완벽하지 않아도 10개 이상의 리스트를 만들 것

**[code]**

In [17]:
from konlpy.tag import Kkma
from collections import Counter

infile = open("소나기.txt", encoding='utf-8')
data=infile.read()

kkma = Kkma()
words=kkma.morphs(data) # 형태소로 구문 분석

vocab=Counter(words)

In [24]:
# 5번 이상인 단어
c=0
for text,count in vocab.items():
    if count >=5:
        print(text,count,end='\t\t')
        c+=1
print("\n--------Total morpheme count is %d------------"%c)

소년 49		은 71		개울가 8		에서 20		소녀 72		를 54		보 25		윤 5		초시 5		네 10		( 5		) 5		이 137		것 35		을 106		알 8		ㄹ 20		수 8		있 54		었 168		다 226		. 311		는 86		개울 6		손 11		고 54		하 77		서 13		지 38		기 15		나 25		ㄴ 78		듯이 6		어 94		, 119		돌아오 7		길 8		에 70		그러 23		데 7		까지 6		더니 6		오늘 5		징검다리 6		한 22		앉 10		아 45		버리 5		가 96		주 5		날 14		게 26		로 16		나오 6		았 41		이번 7		물 12		속 15		ㄴ다 43		얼굴 8		라도 5		움키 5		내 8		는지 6		모르 9		만 12		건너 5		다가 6		하나 8		조약돌 5		그리 7		고는 8		며 15		" 106		바보 7		저 14		도 32		머리 6		달리 8		갈밭 5		가을 5		갈꽃 7		뿐 5		다고 5		생각 13		되 13		않 22		쪽 5		안 8		의 35		같 16		뵈 7		덜 5		집 7		주머니 5		일 6		ㅁ 5		잡 6		전 9		더 17		두 6		으로 27		번 5		말 12		느냐 5		' 12		어서 9		없 8		없이 5		소리 6		오 12		니 9		보이 5		못 6		듣 10		이것 8		? 16		자기 6		눈 6		참 8		산 6		우리 5		멀 5		면 5		허수아비 6		들 7		대 5		달려가 5		그 14		ㄴ다는 5		곳 6		밑 6		수숫단 5		꽃 14		…… 13		그렇 5		크 5		가지 5		송아지 6		비 7		이사 6		호두 6		아버지 5		
--------Total morpheme count is 143------------


In [19]:
# 불용어 리스트
stop_words = ['\ufeff', '가', '을', '는', '것', '이', '·', '의', '다', '.', '\'', '에', '를', ',', '은',
              '크','ㄴ', 'ㄹ', '도', '“', '”', '며', '기', '저', '아', '로', '서', '%', '수', '그', '데',
              '와', '[', ':', ']', 'ㅁ','다고','느냐','니','고','있','었','듯이','……','(',')','\"','ㄴ다는','?','ㄴ다','고는','에서','니']

In [4]:
# 5번 이상인 단어 중 불용어 제외
c=0
for text,count in vocab.items():
    if text not in stop_words and count >=5:
        print(text,count,end='\t\t\t')
        c+=1
print("\n--------Total morpheme count is %d------------"%c)

소년 49		개울가 8		소녀 72		보 25		윤 5		초시 5		네 10		알 8		개울 6		손 11		하 77		지 38		나 25		어 94		돌아오 7		길 8		그러 23		까지 6		더니 6		오늘 5		징검다리 6		한 22		앉 10		버리 5		주 5		날 14		게 26		나오 6		았 41		이번 7		물 12		속 15		얼굴 8		라도 5		움키 5		내 8		는지 6		모르 9		만 12		건너 5		다가 6		하나 8		조약돌 5		그리 7		바보 7		머리 6		달리 8		갈밭 5		가을 5		갈꽃 7		뿐 5		생각 13		되 13		않 22		쪽 5		안 8		같 16		뵈 7		덜 5		집 7		주머니 5		일 6		잡 6		전 9		더 17		두 6		으로 27		번 5		말 12		어서 9		없 8		없이 5		소리 6		오 12		보이 5		못 6		듣 10		이것 8		자기 6		눈 6		참 8		산 6		우리 5		멀 5		면 5		허수아비 6		들 7		대 5		달려가 5		곳 6		밑 6		수숫단 5		꽃 14		그렇 5		가지 5		송아지 6		비 7		이사 6		호두 6		아버지 5		
--------Total morpheme count is 100------------


**[결과]**

Kkma는 지능형 데이터 시스템 (IDS)에서 개발 한 Java로 작성된 형태 학적 분석기 및 자연어 처리 시스템이다.  

Kkma의 morphs 함수는 구문을 형태소로 구문 분석하는 기능을 한다.  

morphs로 분석된 단어들을 collections 모듈의 Counter를 이용해서 각 단어의 숫자를 센다.  

먼저 그 중 5개 이상의 단어들을 출력하고 다음으로 그 중 불용어 리스트에 해당하는 단어를 제거하고 5개 이상의 단어를 다시 출력한다.  

총 단어 중 5개 이상의 단어는 143개이고 43개의 불용어를 53개를 중 5번 이상인 43개가 제거된 100개의 단어를 출력했다.  